pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,511 @@
+package awskms
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"github.com/spiffe/spire/pkg/server/plugin/keymanager"
+	"github.com/spiffe/spire/proto/spire/common/plugin"
+	"github.com/zeebo/errs"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	pluginName       = "awskms"
+	aliasPrefix      = "alias/"
+	defaultKeyPrefix = "SPIRE_SERVER_KEY/"
+
+	keyIDTag = "key_id"
+	aliasTag = "alias"
+)
+
+var (
+	kmsErr = errs.Class(pluginName)
+)
+
+func BuiltIn() catalog.Plugin {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.Plugin {
+	return catalog.MakePlugin(pluginName, keymanager.PluginServer(p))
+}
+
+type keyEntry struct {
+	KMSKeyID  string
+	Alias     string
+	PublicKey *keymanager.PublicKey
+}
+
+// Plugin is the main representation of this keymanager plugin
+type Plugin struct {
+	keymanager.UnsafeKeyManagerServer
+	log       hclog.Logger
+	mu        sync.RWMutex
+	entries   map[string]keyEntry
+	kmsClient kmsClient
+	keyPrefix string
+
+	hooks struct {
+		newClient func(config *Config) (kmsClient, error)
+	}
+}
+
+// Config provides configuration context for the plugin
+type Config struct {
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	Region          string `hcl:"region" json:"region"`
+	KeyPrefix       string `hcl:"key_prefix" json:"key_prefix"`
+}
+
+// New returns an instantiated plugin
+func New() *Plugin {
+	return newPlugin(newKMSClient)
+}
+
+func newPlugin(newClient func(config *Config) (kmsClient, error)) *Plugin {
+	p := &Plugin{}
+	p.hooks.newClient = newClient
+	p.entries = make(map[string]keyEntry)
+	return p
+}
+
+//SetLogger sets a logger
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure sets up the plugin
+func (p *Plugin) Configure(ctx context.Context, req *plugin.ConfigureRequest) (*plugin.ConfigureResponse, error) {
+	config, err := p.validateConfig(req.Configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	p.keyPrefix = config.KeyPrefix
```

Comment:
I changed the code so `Configure` and `GenerateKey` use `Lock()`, since those change the state. Also `SignData`, `GetPublicKey` and `GetPublicKey`to use `RLock()`.